### ISSUE TITLE
try with resources with Java 11 triggers RCN warnings

### ISSUE THREAD

**Issue by ijuma** [spotbugs/spotbugs#756]
Tested with spotBugs 3.1.7. The issue doesn't happen with Java 10.

Example:

> RCN | Nullcheck of fc at line 629 of value previously dereferenced in org.apache.kafka.common.utils.Utils.readFileAsString(String, Charset)

```java
public static String readFileAsString(String path, Charset charset) throws IOException {
  try (FileChannel fc = FileChannel.open(Paths.get(path))) {
    MappedByteBuffer bb = fc.map(FileChannel.MapMode.READ_ONLY, 0, fc.size()); // line 629
    return StandardCharsets.UTF_8.decode(bb).toString();
  }
}
```

**Comment by muminc** [comment]
This Jacoco issue https://github.com/jacoco/jacoco/pull/669 says Java 11 compiler is producing different byte code pattern for TRY with Resources.

I'm not familiar with byte code.  Anyone know if this is Compiler issue or SpotBugs issue ?

**Comment by ijuma** [comment]
Looking at the bytecode for:

```java
    public static void readFileAsString(String path) throws IOException {
        try (FileChannel fc = FileChannel.open(Paths.get(path))) {
            fc.size();
        }

    }
```

Bytecode:

```java
  public static void readFileAsString(java.lang.String) throws java.io.IOException;
    Code:
       0: aload_0
       1: iconst_0
       2: anewarray     #2                  // class java/lang/String
       5: invokestatic  #3                  // Method java/nio/file/Paths.get:(Ljava/lang/String;[Ljava/lang/String;)Ljava/nio/file/Path;
       8: iconst_0
       9: anewarray     #4                  // class java/nio/file/OpenOption
      12: invokestatic  #5                  // Method java/nio/channels/FileChannel.open:(Ljava/nio/file/Path;[Ljava/nio/file/OpenOption;)Ljava/nio/channels/FileChannel;
      15: astore_1
      16: aload_1
      17: invokevirtual #6                  // Method java/nio/channels/FileChannel.size:()J
      20: pop2
      21: aload_1
      22: ifnull        52
      25: aload_1
      26: invokevirtual #7                  // Method java/nio/channels/FileChannel.close:()V
      29: goto          52
      32: astore_2
      33: aload_1
      34: ifnull        50
      37: aload_1
      38: invokevirtual #7                  // Method java/nio/channels/FileChannel.close:()V
      41: goto          50
      44: astore_3
      45: aload_2
      46: aload_3
      47: invokevirtual #9                  // Method java/lang/Throwable.addSuppressed:(Ljava/lang/Throwable;)V
      50: aload_2
      51: athrow
      52: return
    Exception table:
       from    to  target type
          16    21    32   Class java/lang/Throwable
          37    41    44   Class java/lang/Throwable
```

It seems like spotBugs is complaining about:

```
      16: aload_1
      17: invokevirtual #6                  // Method java/nio/channels/FileChannel.size:()J
      20: pop2
      21: aload_1
      22: ifnull        52
      25: aload_1
      26: invokevirtual #7                  // Method java/nio/channels/FileChannel.close:()V
```

We invoke a method on `FileChannel` and then check if it's null. It seems like that should not be strictly required.

**Comment by ijuma** [comment]
Based on the following (from the javac code), it seems like the "provably non-null" check is not being triggered for the  catch block.

```java
     *   try ResourceSpecificationtail
     *     Block
     *   } body-only-finally {
     *     if (#resource != null) //nullcheck skipped if Expression is provably non-null
     *         #resource.close();
     *   } catch (Throwable #primaryException) {
     *       if (#resource != null) //nullcheck skipped if Expression is provably non-null
     *           try {
     *               #resource.close();
     *           } catch (Throwable #suppressedException) {
     *              #primaryException.addSuppressed(#suppressedException);
     *           }
     *       throw #primaryException;
     *   }
     * }
```

**Comment by iloveeclipse** [comment]
If I read it right, the first null check is not needed here, because it can't be reached with null - either we have the resource or exception

**Comment by ijuma** [comment]
@iloveeclipse that's what it looks to me too. It's unlikely that this will be fixed in javac for the final release of Java 11 though. Is a workaround possible in spotBugs or do we have to wait for a potential fix in javac?

**Comment by iloveeclipse** [comment]
Hmm. How one should know if this code is generated and not written by humans... In theory everything is possible, but buggy code is buggy.

**Comment by ijuma** [comment]
I emailed the javac mailing list: http://mail.openjdk.java.net/pipermail/compiler-dev/2018-September/012426.html

**Comment by ijuma** [comment]
@eolivelli maybe you can reply to my OpenJDK mailing list in case that increases the chance of a response. :)

**Comment by ctmay4** [comment]
Would it be better to submit an issue to the bug tracker?

https://bugs.openjdk.java.net/projects/JDK/issues/JDK-8213451

**Comment by iloveeclipse** [comment]
> Would it be better to submit an issue to the bug tracker?

Sure, just try to get an account there :-)

Long time ago I was told that you have to write to the mailing list, and someone will pick your message up and create a bug. But the mails on that list were all unrelated, spammed my account and it wasn't something for me. So I gave up. If anyone on this list has more time, it would be really helpful to get it into the official OpenJDK tracker.

**Comment by eolivelli** [comment]
@ijuma 

> @eolivelli maybe you can reply to my OpenJDK mailing list in case that increases the chance of a response. :)

I did not receive that email, so I can't reply to it, is there any way to reply to that email without having received it ?
I have now subscribed my address to compiler-dev

**Comment by ctmay4** [comment]
> > Would it be better to submit an issue to the bug tracker?
> 
> Sure, just try to get an account there :-)
> 
> Long time ago I was told that you have to write to the mailing list, and someone will pick your message up and create a bug. But the mails on that list were all unrelated, spammed my account and it wasn't something for me. So I gave up. If anyone on this list has more time, it would be really helpful to get it into the official OpenJDK tracker.

Oops.  I've never tried.  I just assumed anyone could make an account.  My bad.

**Comment by iloveeclipse** [comment]
> I just assumed anyone could make an account

It is **Open**JDK but it is **Closed**JDK bug tracker. It is really pity.

**Comment by eolivelli** [comment]
@ijuma thank you.
I sent a reply to your email just now

**Comment by ijuma** [comment]
There have been a few responses in the OpenJDK mailing list and they said that javac is doing the right thing.

> On 11/9/2018 12:31 PM, Vicente Romero wrote:
> > My evaluation on this issue is that javac is generating code
> > according to the spec, see JLS11 14.20.3.1 [1]. There is an explicit
> > check for null on the resource before invoking the close() method.
> > This is the null check javac is generating.
> 
> I confirm that the explicit null check on the resource variable is 
> required, per the translation in 
> https://docs.oracle.com/javase/specs/jls/se11/html/jls-14.html#jls-14.20.3.1-140 
> -- it has been specified that way since JLS7.
> 
> > Fix for JDK-8194978 went a bit further trying to eliminate dead code
> > for some cases for which javac could statically determine that the
> > resource would be different from null. In particular if the resource
> > was initialized with a new class expression.
> 
> I confirm it's acceptable to not perform the null check in generated 
> code if you can prove the resource variable is non-null. Frankly, 
> though, there is so much that can occur in a "new class expression" -- 
> `new Outer(a(b(c)))<>.new <String>Inner(d(e(f)))` -- that I would be 
> wary of trying to prove anything.
> 
> > It could be argued that we could try to analyze the body and if we
> > find out that a NPE must be thrown in the body of the try, then the
> > null check on the resource would be deemed unnecessary. But then we
> > can get to an implementation that will be out of sync with the spec
> > plus it probably won't cover all cases.
> 
> Right, let's not be too clever.

**Comment by eolivelli** [comment]
Is there any smart way to disable this rule only on JDK11+ ?

I am using Maven (but for Gradle I guess it is the same problem), and I will have to add two different exclusion filter depending on the JDK in use, it is very tricky and error prone.

If there is no way to address this issue, because spotbugs is telling the truth, can we disable the rule on JDK11+ ? it is a pity but it is also a problem to fill every project of "ignore this rule"

**Comment by lajosrodek** [comment]
Looks like this is duplicated by #600 and #878, so this one should probably get labeled with [false positive](/spotbugs/spotbugs/labels/false%20positive) and [Java-11](/spotbugs/spotbugs/labels/Java-11).

**Comment by mmoehring** [comment]
Same on Java 12, if you want to add that label.

**Comment by efenderbosch** [comment]
Also seeing this w/ JDK 8.

java -version
openjdk version "1.8.0_212"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_212-b03)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.212-b03, mixed mode)

**Comment by ccrvincent** [comment]
I believe this is a duplicate of #259 , isn't it?

**Comment by stevenschlansker** [comment]
Still a problem with Java 13 :(
If there's no fix coming, this rule should probably be disabled, or at least ignore the known-bad construction.

**Comment by C-Otto** [comment]
FYI: Problem exists with Java 14.

**Comment by eolivelli** [comment]
Yep, if no one fixes it it will stay.

If anyone could help with this fix....it will be a great help to the whole user base of Spotbugs and derived products

**Comment by mepeisen** [comment]
I did not read the openjdk mailing list. But I played around a bit and refactored everything to methods.
I ended up with following working code:
`try (MyResource some = createMyResource()) {
…
}

private @Nonnull InputStream createMyResource() {
…
}`

The important Change is that createResource guarantees the returned value being non null through Annotation javax.annotation.Nonnull.
Can you verify this "fix"? Can anyone look at the bytecode?
However current Spotbugs does no longer complain. For me this seems to be clear. As createMyResource guarantees returning nunnull the compiler seems to skip the wrong null checks.

**Comment by KengoTODA** [comment]
closed by #1248

**Comment by stefanos-kalantzis** [comment]
any clues when this will become available in maven central? or is there some other way of obtaining it in our maven build?
(without building it ourselves)

**Comment by koalalam** [comment]
Got the same issue with Spotbugs 4.4.2:

Mockito (org.mockito:mockito-inline) version: 4.0.0

```java
import static org.mockito.Mockito.mockStatic;

try (MockedStatic<SomeStaticUtilClass> mock = mockStatic(SomeStaticUtilClass.class)) {
...
}
```

**Comment by oehme** [comment]
Same with `Corretto-11.0.10.9.1` and Spotbugs 4.4.2, this issue does not seem to be fixed.

**Comment by nedtwigg** [comment]
Still present in `4.5.0`.

**Comment by colin-janke** [comment]
It seems this issue is closed but #1338 and #1694 are continuing it.

**Comment by ctubbsii** [comment]
In what version of spotbugs was this fixed? There is no milestone marked on any of the issues/PRs in GitHub to make that easy to determine.
